src/trg-client.c

```diff
@@ -480,360 +430,272 @@ void trg_client_configunlock(TrgClient *tc)
+    }
 
-    if (decode_error) {
-        g_error("JSON decoding error: %s", decode_error->message);
-        g_error_free(decode_error);
-        response->status = FAIL_JSON_DECODE;
-        return response;
+    root = json_parser_steal_root(parser);
+    if (!root) {
+        status = FAIL_JSON_DECODE;
+        goto out;
     }
 
-    result = json_object_get_member(response->obj, FIELD_RESULT);
-    if (!result || g_strcmp0(json_node_get_string(result), FIELD_SUCCESS))
-        response->status = FAIL_RESPONSE_UNSUCCESSFUL;
+    obj = json_node_dup_object(root);
+    json_object_seal(obj);
 
-    return response;
+    rpc_result = json_object_get_member(obj, FIELD_RESULT);
+    if (!rpc_result || g_strcmp0(json_node_get_string(rpc_result), FIELD_SUCCESS))
+        status = FAIL_RESULT_UNSUCCESSFUL;
+
+out:
+    trg_request_callback(request, obj, status, err_msg);
 }
 
-trg_response *dispatch_public_http(TrgClient *tc, trg_request *req)
+static void rpc_callback(GObject *source, GAsyncResult *result, gpointer user_data)
 {
-    trg_response *response = g_new0(trg_response, 1);
-    CURL *curl = get_curl(tc, HTTP_CLASS_PUBLIC);
-    struct curl_slist *headers = NULL;
-    long httpCode = 0;
-    gchar *cookie_header = NULL;
-
-    response->size = 0;
-    response->raw = NULL;
+    g_autoptr(GError) error = NULL;
+    g_autoptr(GInputStream) stream = soup_session_send_finish(SOUP_SESSION(source), result, &error);
+    trg_request *request = user_data;
 
-    curl_easy_setopt(curl, CURLOPT_URL, req->url);
-    curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)response);
+    if (error) {
+        if (g_error_matches(error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+            return;
 
-    if (req->cookie) {
-        cookie_header = g_strdup_printf("Cookie: %s", req->cookie);
-        headers = curl_slist_append(NULL, cookie_header);
+        trg_request_callback(request, NULL, FAIL_HTTP_UNSUCCESSFUL, g_strdup(error->message));
+        return;
     }
 
-    if (headers)
-        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
-
-    response->status = curl_easy_perform(curl);
-
-    trg_request_free(req);
-
-    g_free(cookie_header);
+    gint status = soup_message_get_status(request->msg);
+    if (status != SOUP_STATUS_OK) {
+        trg_request_callback(request, NULL, status, NULL);
+        return;
+    }
 
-    g_clear_pointer(&headers, curl_slist_free_all);
+    g_autoptr(JsonParser) parser = json_parser_new();
+    json_parser_load_from_stream_async(parser, stream, NULL, json_callback, request);
+}
 
-    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &httpCode);
+static gboolean tls_callback(SoupMessage *msg, GTlsCertificate *cert,
+                             GTlsCertificateFlags tls_errors, gpointer user_data)
+{
+    trg_request *request = user_data;
 
-    if (response->status == CURLE_OK && httpCode != HTTP_OK) {
-        response->status = (-httpCode) - 100;
-    }
+    if (trg_client_get_ssl_validate(request->client))
```

Comment:
If validation is enabled I believe you should always return `FALSE`. The reason this callback is called because it was invalid.